test/e2e/sequence_test.go

```diff
@@ -0,0 +1,138 @@
+// +build e2e
+
+/*
+Copyright 2019 The Knative Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	eventingv1alpha1 "github.com/knative/eventing/pkg/apis/eventing/v1alpha1"
+	messagingv1alpha1 "github.com/knative/eventing/pkg/apis/messaging/v1alpha1"
+	"github.com/knative/eventing/test/base/resources"
+	"github.com/knative/eventing/test/common"
+	"k8s.io/apimachinery/pkg/util/uuid"
+)
+
+func TestSequence(t *testing.T) {
+	const (
+		sequenceName  = "e2e-sequence"
+		senderPodName = "e2e-sequence-sender-pod"
+
+		channelName      = "e2e-sequence-channel"
+		subscriptionName = "e2e-sequence-subscription"
+		loggerPodName    = "e2e-sequence-logger-pod"
+	)
+	stepSubscriberConfigs := []struct {
+		podName     string
+		msgAppender string
+	}{{
+		podName:     "e2e-stepper1",
+		msgAppender: "-step1",
+	}, {
+		podName:     "e2e-stepper2",
+		msgAppender: "-step2",
+	}, {
+		podName:     "e2e-stepper3",
+		msgAppender: "-step3",
+	}}
+	channelTypeMeta := common.InMemoryChannelTypeMeta
+
+	client := setup(t, true)
+	defer tearDown(client)
+
+	// construct steps for the sequence
+	steps := make([]eventingv1alpha1.SubscriberSpec, 0)
+	for _, config := range stepSubscriberConfigs {
+		// create a stepper Pod with Service
+		podName := config.podName
+		msgAppender := config.msgAppender
+		stepperPod := resources.SequenceStepperPod(podName, msgAppender)
+
+		client.CreatePodOrFail(stepperPod, common.WithService(podName))
+		// create a new step
+		step := eventingv1alpha1.SubscriberSpec{
+			Ref: resources.ServiceRef(podName),
+		}
+		// add the step into steps
+		steps = append(steps, step)
+	}
+
+	// create channelTemplate for the Sequence
+	channelTemplate := messagingv1alpha1.ChannelTemplateSpec{
+		TypeMeta: *(channelTypeMeta),
+	}
+
+	// create channel as reply of the Sequence
+	// TODO(Fredy-Z): now we'll have to use a channel plus its subscription here, as reply of the Sequence
+	//                must be Addressable. In the future if we use Knative Serving in the tests, we can
+	//                make the logger service as a Knative service, and remove the channel and subscription.
```

Comment:
I thought we could use a kubernetes Service as a subscriber. Is that wrong?

Reply:
`Reply` of the `Sequence` must be `Addressable`, since `Sequence` will create `Subscriptions` for its `steps`, and the last `Subscription` will use it as its `Reply`, which must be `Addressable`...
`Subscriber` of `Subscription` can be a Kubernetes Service.

I think it's expected behavior(?), but it does add some confusion here.

Reply:
The spec mentioned `Reply` of `Subscription` must be a `Channel` - https://github.com/knative/eventing/blob/master/docs/spec/spec.md, but seems we don't have such a restriction in code.